client/pkg/transport/listener_test.go

```diff
@@ -573,3 +578,140 @@ func TestSocktOptsEmpty(t *testing.T) {
+			expectHandshakeError: false,
+		},
+		"client cert is revoked": {
+			expectHandshakeError: true,
+			revokedCertificates: []pkix.RevokedCertificate{
+				{
+					SerialNumber:   clientCert.SerialNumber,
+					RevocationTime: time.Now(),
+				},
+			},
+		},
+		"invalid CRL file content": {
+			expectHandshakeError:   true,
+			revocationListContents: []byte("@invalidcontent"),
+		},
+	}
+
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			tmpdir := t.TempDir()
+			tlsInfo, err := createSelfCert(t)
+			if err != nil {
+				t.Fatalf("unable to create server cert: %v", err)
+			}
+			tlsInfo.TrustedCAFile = clientTLSInfo.CertFile
+			tlsInfo.CRLFile = filepath.Join(tmpdir, "revoked.r0")
+
+			loaded, err := os.ReadFile(tlsInfo.CertFile)
+			if err != nil {
+				t.Fatalf("unable to read server cert: %v", err)
+			}
+			block, _ := pem.Decode(loaded)
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				t.Fatalf("unable to decode server cert: %v", err)
+			}
+
+			loaded, err = os.ReadFile(tlsInfo.KeyFile)
+			if err != nil {
+				t.Fatalf("unable to read server key: %v", err)
+			}
+			block, _ = pem.Decode(loaded)
+			key, err := x509.ParseECPrivateKey(block.Bytes)
+			if err != nil {
+				t.Fatalf("unable to parse server key: %v", err)
+			}
+
+			revocationListContents := test.revocationListContents
+			if len(revocationListContents) == 0 {
+				tmpl := &x509.RevocationList{
+					RevokedCertificates: test.revokedCertificates,
+					ThisUpdate:          time.Now(),
+					NextUpdate:          time.Now().Add(time.Hour),
+					Number:              big.NewInt(1),
+				}
+				revocationListContents, err = x509.CreateRevocationList(rand.Reader, tmpl, cert, key)
+				if err != nil {
+					t.Fatalf("unable to create revocation list: %v", err)
+				}
+			}
+
+			if err = os.WriteFile(tlsInfo.CRLFile, revocationListContents, 0600); err != nil {
+				t.Fatalf("unable to write revocation list: %v", err)
+			}
+
+			chHandshakeFailure := make(chan error, 1)
+			tlsInfo.HandshakeFailure = func(_ *tls.Conn, err error) {
+				if err != nil {
+					chHandshakeFailure <- err
+				}
+			}
+
+			rootCAs := x509.NewCertPool()
+			rootCAs.AddCert(cert)
+
+			clientCert, err := tls.LoadX509KeyPair(clientTLSInfo.CertFile, clientTLSInfo.KeyFile)
+			if err != nil {
+				t.Fatalf("unable to create peer cert: %v", err)
+			}
+
+			ln, err := NewListener("127.0.0.1:0", "https", tlsInfo)
+			if err != nil {
+				t.Fatalf("unable to start listener: %v", err)
+			}
+			defer ln.Close()
+
+			tlsConfig := &tls.Config{}
+			tlsConfig.InsecureSkipVerify = false
+			tlsConfig.Certificates = []tls.Certificate{clientCert}
+			tlsConfig.RootCAs = rootCAs
+
+			tr := &http.Transport{TLSClientConfig: tlsConfig}
+			cli := &http.Client{Transport: tr}
+			go cli.Get("https://" + ln.Addr().String())
+
+			chAcceptConn := make(chan net.Conn, 1)
+			go func() {
+				conn, _ := ln.Accept()
+				chAcceptConn <- conn
+			}()
```

Comment:
Suggest to use a WaitGroup to ensure both goroutines finish before exiting this test,

Reply:
Hey @ahrtr, I tried adding a wait group, but the code got more complex as I will need to manually close the listener from line 684, as this goroutine has a , and the connection from this routine so the wait group condition meets. If you feel that would be the right way, I can push that change. The implementation is this one:

```go
			ln, err := NewListener("127.0.0.1:0", "https", tlsInfo)
			if err != nil {
				t.Fatalf("unable to start listener: %v", err)
			}

			tlsConfig := &tls.Config{}
			tlsConfig.InsecureSkipVerify = false
			tlsConfig.Certificates = []tls.Certificate{clientCert}
			tlsConfig.RootCAs = rootCAs

			tr := &http.Transport{TLSClientConfig: tlsConfig}
			cli := &http.Client{Transport: tr}
			var wg sync.WaitGroup
			wg.Add(2)
			go func() {
				defer wg.Done()
				cli.Get("https://" + ln.Addr().String())
			}()

			chAcceptConn := make(chan net.Conn, 1)
			go func() {
				defer wg.Done()
				conn, err := ln.Accept()
				if err == nil {
					chAcceptConn <- conn
				}
			}()

			select {
			case err := <-chHandshakeFailure:
				if !test.expectHandshakeError {
					t.Errorf("expecting no handshake error, got: %v", err)
				}
			case conn := <-chAcceptConn:
				if test.expectHandshakeError {
					t.Errorf("expecting handshake error, got nothing")
				}
				conn.Close()
			}
			ln.Close()
			wg.Wait()
```

vs (using defer to close the listener and the connection)

```go
			ln, err := NewListener("127.0.0.1:0", "https", tlsInfo)
			defer ln.Close()
			if err != nil {
				t.Fatalf("unable to start listener: %v", err)
			}

			tlsConfig := &tls.Config{}
			tlsConfig.InsecureSkipVerify = false
			tlsConfig.Certificates = []tls.Certificate{clientCert}
			tlsConfig.RootCAs = rootCAs

			tr := &http.Transport{TLSClientConfig: tlsConfig}
			cli := &http.Client{Transport: tr}
			go cli.Get("https://" + ln.Addr().String())

			chAcceptConn := make(chan net.Conn, 1)
			go func() {
				conn, err := ln.Accept()
				if err == nil {
					chAcceptConn <- conn
				}
			}()

			select {
			case err := <-chHandshakeFailure:
				if !test.expectHandshakeError {
					t.Errorf("expecting no handshake error, got: %v", err)
				}
			case conn := <-chAcceptConn:
				defer conn.Close()
				if test.expectHandshakeError {
					t.Errorf("expecting handshake error, got nothing")
				}
			}

```

Reply:
The first one makes more sense to me. We need to try to ensure everything under control and work as expected.